src/cosl/coordinated_workers/nginx.py

```diff
@@ -29,6 +181,478 @@
+    """If True, overrides `worker_role` and routes to all available backend servers.
+
+    Use this when the upstream should be generic and include any available backend.
+    """
+
+
+class NginxConfig:
+    """Responsible for building the Nginx configuration used by the coordinators."""
+
+    _worker_processes = "5"
+    _pid = "/tmp/nginx.pid"
+    _worker_rlimit_nofile = "8192"
+    _worker_connections = "4096"
+    _proxy_read_timeout = "300"
+    _supported_tls_versions = ["TLSv1", "TLSv1.1", "TLSv1.2", "TLSv1.3"]
+    _ssl_ciphers = ["HIGH:!aNULL:!MD5"]
+    _proxy_connect_timeout = "5s"
+
+    def __init__(
+        self,
+        server_name: str,
+        upstream_configs: List[NginxUpstream],
+        server_ports_to_locations: Dict[int, List[NginxLocationConfig]],
+        enable_health_check: bool = False,
+        enable_status_page: bool = False,
+    ):
+        """Constructor for an Nginx config generator object.
+
+        Args:
+            server_name: The name of the server (e.g. coordinator fqdn), which is used to identify the server in Nginx configurations.
+            upstream_configs: List of Nginx upstream metadata configurations used to generate Nginx `upstream` blocks.
+            server_ports_to_locations: Mapping from server ports to a list of Nginx location configurations.
+            enable_health_check: If True, adds a `/` location that returns a basic 200 OK response.
+            enable_status_page: If True, adds a `/status` location that enables `stub_status` for basic Nginx metrics.
+
+        Example:
+            .. code-block:: python
+            NginxConfig(
+            server_name = "tempo-coordinator-0.tempo-coordinator-endpoints.model.svc.cluster.local",
+            upstreams = [
+                NginxUpstream(name="zipkin", port=9411, worker_role="distributor"),
+            ],
+            server_ports_to_locations = {
+                9411: [
+                    NginxLocationConfig(
+                        path="/",
+                        backend="zipkin"
+                    )
+                ]
+            })
+        """
+        try:
+            # we lazy-load crossplane to avoid adding a dependency to cosl as a whole
+            import crossplane  # type: ignore
+
+            self._builder = crossplane.build  # type: ignore
+        except ModuleNotFoundError:
+            raise RuntimeError(
+                "Unmet dependency: the coordinated-workers NginxConfig object depends on the 'crossplane' package. \
+            Please install it with: `pip install crossplane` and add it to your charm's dependencies."
+            )
+        self._server_name = server_name
+        self._dns_IP_address = self._get_dns_ip_address()
+        self._ipv6_enabled = is_ipv6_enabled()
+        self._upstream_configs = upstream_configs
+        self._server_ports_to_locations = server_ports_to_locations
+        self._enable_health_check = enable_health_check
```

Comment:
very minor nit, but I tend to try and do all 'init-arg-to-attr' assignments at the top of init, before any method calls, function calls, or other 'derived' attribute initialization. that way it's clearer to read where the init arguments are used.